[AutoDiff] Canonicalize linear maps to be maximally indirect.

dan-zheng opened the pull request:
Canonicalize JVPs/VJPs to return maximally abstracted linear map functions
with `@in_guaranteed` parameters and an `@out` result.

This is a necessary step towards re-enabling LoadableByAddress:
linear map type is no longer computed based on the original function type.

Summary of changes:
- SILGen
  - Extra reabstraction logic for JVP/VJPs and differential/pullbacks to
    handle maximally indirect linear maps.
  - [Related forum discussion.](https://forums.swift.org/t/how-to-best-maximally-reabstract-returned-function-type-in-silgen/27945)
- Differentiation transform
  - Use only tangent/adjoint buffers in differential/pullback generation.

See [TF-11](https://bugs.swift.org/projects/TF/issues/TF-11) for more info regarding LoadableByAddress.
See [TF-625](https://bugs.swift.org/projects/TF/issues/TF-625) for more info regarding maximally abstracted linear maps.

---

Initial patch by @rxwei - commits lost because resolving conflicts got too complicated, sorry.

dan-zheng committed:
[AutoDiff] Canonicalize linear maps to be maximally indirect.

Canonicalize JVPs/VJPs to return maximally abstracted linear map functions
with `@in_guaranteed` parameters and an `@out` result.

This is a necessary step towards re-enabling LoadableByAddress:
linear map type is no longer computed based on the original function type.

See TF-11 for more info regarding LoadableByAddress.
See TF-625 for more info regarding maximally abstracted linear maps.

dan-zheng commented:
Current status:

Standard library compilation passes. All tests pass except temporary memory leak regressions, to be fixed before merging:
```
Failing Tests (2):
    Swift(macosx-x86_64) :: AutoDiff/leakchecking.swift
    Swift(macosx-x86_64) :: AutoDiff/silgen_thunking/main.swift
```

Re-enabling `LoadableByAddress` works, except for one case - fix will be investigated in a follow-up.

dan-zheng committed:
Merge remote-tracking branch 'apple/tensorflow' into ad-all-indirect

dan-zheng committed:
Merge remote-tracking branch 'apple/tensorflow' into ad-all-indirect

dan-zheng committed:
Merge branch 'tensorflow' of github.com:apple/swift into ad-all-indirect

dan-zheng committed:
Progress: all tests are fixed.

- Change differential generation to use only tangent buffers.
- Change pullback visitors to use only adjoint buffers.
- Change adjoint of active value propagation to use only adjoint buffers.
- Mark all tangent/adjoint value helpers as `[[deprecated]]`.
  - They are not deleted because helpers may become useful after
    SIL opaque values are introduced.

dan-zheng committed:
Add todo/explanatory comments.

dan-zheng committed:
Undo changes to LoadableByAddress.

Will re-add in follow-up for separation of concerns.

dan-zheng committed:
Clean up.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
It would be good to test the performance impact of this patch. Using adjoint buffers instead of adjoint values may incur a performance penalty.

rxwei commented on lib/SIL/SILFunctionType.cpp:
```diff
@@ -2496,6 +2456,46 @@ static CanType copyOptionalityFromDerivedToBase(TypeConverter &tc,
 
+// SWIFT_ENABLE_TENSORFLOW
+// TODO: Move this to a common location (SILFunctionType.cpp).
+static CanSILFunctionType
```
Add a comment. What does “normalize” mean?

dan-zheng replied:
Added comment in https://github.com/apple/swift/pull/26709/commits/1093416710d66105e8dc09a6110aed839b535975.

rxwei commented on lib/SIL/SILFunctionType.cpp:
```diff
@@ -2496,6 +2456,46 @@ static CanType copyOptionalityFromDerivedToBase(TypeConverter &tc,
+}
+
+// SWIFT_ENABLE_TENSORFLOW
+// TODO: Move this to a common location (SILFunctionType.cpp).
```
Add a comment.

dan-zheng replied:
Added comment in https://github.com/apple/swift/pull/26709/commits/1093416710d66105e8dc09a6110aed839b535975.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -7433,6 +7317,7 @@ AdjointValue PullbackEmitter::makeAggregateAdjointValue(
   return AdjointValue::createAggregate(allocator, remapType(type), elements);
 }
 
+[[deprecated]]
```
I’d suggest removing all deprecated unused functions because they are going to be untested from now on.

dan-zheng replied:
How about deleting all `AdjointValue` construction/materialization/accumulation utilities, except for the definition of `AdjointValue `itself? `AdjointValue` itself can be repurposed for `AdjointBuffer` symbolic zero optimization.

dan-zheng replied:
Done in https://github.com/apple/swift/pull/26709/commits/82c775f10a1dadc41df702e31fa4d47d0f6c6469. Minus 500+ lines.

rxwei replied:
That sounds fine. We'll probably want to rename it to `SymbolicTangent` or something.

dan-zheng committed:
Remove unused `AdjointValue` utilities.

Remove unused `AdjointValue` utilities for materialization and addition.
`AdjointValue` itself is preserved to implement symbolic zero buffer optimization.

dan-zheng committed:
Add doc comments for autodiff function type normalization.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -5206,30 +5128,41 @@ class JVPEmitter final
+      // should not be overwritten.
+      diffBuilder.setInsertionPoint(
+          differential.getEntryBlock(),
+           getNextDifferentialLocalAllocationInsertionPoint());
```
```suggestion
          getNextDifferentialLocalAllocationInsertionPoint());
```

dan-zheng replied:
Fixed in https://github.com/apple/swift/pull/26709/commits/bc2657784cb232a2f85921b8446883762a079f18.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -6119,29 +5924,48 @@ class PullbackEmitter final : public SILInstructionVisitor<PullbackEmitter> {
-    assert(originalBuffer->getType().isAddress());
-    assert(originalBuffer->getFunction() == &getOriginal());
-    auto insertion = bufferMap.try_emplace({origBB, originalBuffer},
+  SILValue makeLocalAllocation(SILLocation loc, SILType type) {
```
I couldn't find a single caller of this method. Dead code?

dan-zheng replied:
Thanks - removed in https://github.com/apple/swift/pull/26709/commits/bc2657784cb232a2f85921b8446883762a079f18.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -7190,44 +6873,37 @@ class PullbackEmitter final : public SILInstructionVisitor<PullbackEmitter> {
+        continue;
+      auto tmpElement = builder.createTupleElementAddr(loc, tmp, adjIdx);
+      if (tei->getFieldNo() == i) {
+        builder.createCopyAddr(
```
This tmp copy is redundant. Instead, just emit `tuple_element_addr` and call `addToAdjointBuffer` with the address to the element.

dan-zheng replied:
I see, that makes sense.

After differentiation is moved before ownership eliminator, I'm not able to find a test case that tests `PullbackEmitter` visitor for `tuple_extract`. The old `tuple_extract` reproducer now generates `destructure_tuple`:

```swift
func foo(_ x: Float, _ y: Float) -> (Float, Float) {
  return (x + 1, y + 2)
}
@_silgen_name("bar")
func bar(_ x: Float, _ y: Float) -> Float {
  // Try to test `tuple_extract` instruction.
  // Instead, generates `destructure_tuple` since AD comes before ownership eliminator.
  foo(x, y).1
}
print((0, 1), gradient(at: 3, 3, in: bar))
```

```
[AD] Original bb0: To differentiate or not to differentiate?
[ ]   return %7 : $Float                              // id: %8
[∂]   (%6, %7) = destructure_tuple %5 : $(Float, Float) // user: %8
[∂]   %5 = apply %4(%0, %1) : $@convention(thin) (Float, Float) -> (Float, Float) // user: %6
[ ]   // function_ref foo(_:_:)
  %4 = function_ref @$s5hello3fooySf_SftSf_SftF : $@convention(thin) (Float, Float) -> (Float, Float) // user: %5
[ ]   debug_value %1 : $Float, let, name "y", argno 2 // id: %3
[ ]   debug_value %0 : $Float, let, name "x", argno 1 // id: %2
```

What should we do with the `tuple_extract` visitor? Perhaps we can write a SIL test for it.

dan-zheng replied:
> What should we do with the tuple_extract visitor? Perhaps we can write a SIL test for it.

Actually, I'm leaning towards deleting `tuple_extract` visitors for differential and pullback generation, since it is not testable.

dan-zheng replied:
Deleted `tuple_extract` visitors in https://github.com/apple/swift/pull/26709/commits/cce599ddc42d6453522b9b1900fb115b8ee8d4db.

dan-zheng replied:
I applied the same redundant `tmp` allocation optimize for `struct_extract` in https://github.com/apple/swift/pull/26709/commits/a6988c8f54ec6858da48a7f1fe2b8608368c6432.

rxwei replied:
If you do a grep, you'll find that SILGen does generate `tuple_extract` sometimes. While it's hard to come up with a Swift code that results in `tuple_extract`, I think it's better to write a SIL FileCheck test case that triggers this code path than removing the code.

dan-zheng replied:
Fair, I'll look into add a `tuple_extract` SIL test today.

dan-zheng committed:
Minor cleanup.

Address review feedback.
Standardize variable naming.

dan-zheng committed:
Delete `tuple_extract` visitors for differential/pullback generation.

`tuple_extract` is no longer testable after Differentiation has been moved
before OwnershipModelEliminator. Old reproducers for `tuple_extract` now
generate `destructure_tuple`.

Deleting untestable code is prudent.

dan-zheng committed:
Optimize `PullbackEmitter::visitStructExtractInst`.

Avoid unnecessary local tangent struct struct allocation.
Generate `struct_element_addr` into existing struct tangent buffer.

dan-zheng committed:
[WIP] Re-add differential/pullback support for `tuple_extract`.

Todo: add test. Currently, test crashes.
This commit will be overwritten and rebased when done.

rxwei committed:
Fix build error by adding a missing `}`.

rxwei committed:
Improve code style around macro definitions.

dan-zheng committed:
Merge branch 'tensorflow' of github.com:apple/swift into ad-all-indirect

dan-zheng committed:
Remove `tuple_extract` visitors.

There are no known tests for `tuple_extract` visitors.
Re-adding `tuple_extract` visitors may be necessary when differentiation
supports inout parameters.

This Gist shows cases where SILGen produces `struct_extract`:
https://gist.github.com/dan-zheng/1343673d2d4d20d403306283b42d522b

dan-zheng committed:
Fix all tests, clean up.

Some cleanup will be refactored to a separate patch.

rxwei committed:
Merge branch 'tensorflow' of github.com:apple/swift into ad-all-indirect

dan-zheng commented:
A less invasive approach for re-enabling LoadableByAddress was found: https://github.com/apple/swift/pull/27923

Maximal indirection will not be pursued further: the differentiation transform should not be forced to generate maximally indirect code to workaround issues caused by LoadableByAddress.
